src/ipath/service.rs

```diff
@@ -1,25 +1,31 @@
 extern crate libc;
 
-use std::os::unix::io::RawFd;
 use std::ffi::CString;
-use fileops::{open, close};
+use fileops::Fd;
 use errno::*;
+use std::os::unix::io::AsRawFd;
 
-fn ipath_context_open(unit: isize) -> Option<libc::c_int> {
+fn ipath_context_open(unit: isize) -> Option<Fd> {
   let dev_path = if unit >= 0 {
     format!("/dev/ipath{}", unit)
   } else {
     format!("/dev/ipath")
   };
 
+  // XXX: Do we need ipath_wait_for_device? it literally just waits.
+
   // open and fcntl return -1 and set errno in the case of an error
   // the fd here is the result, but we need to check fcntl's result
-  let fd = open(dev_path, libc::O_RDWR);
-  // We can get away w/o checking open, because fcntl on fd -1 does nothing
-  if unsafe { libc::fcntl(fd, libc::F_SETFD, libc::FD_CLOEXEC) } >= 0 {
-    Some(fd)
-  } else {
-    None
+  let fd_maybe = Fd::open(dev_path, libc::O_RDWR);
+  match fd_maybe {
+    Some(fd) => {
+      if fd.fcntl(libc::F_SETFD, libc::FD_CLOEXEC).is_some() {
```

Comment:
What are we checking here? This interaction seems a bit idiomatic.

So I've maybe got a thing, "fd". I'm calling a function on it that return an option, and if that option is there, I keep my fd. So, our contract is if our "fd" is present, it's fnctl value is also present?